src/tools/cephfs/first-damage.py

```diff
@@ -0,0 +1,137 @@
+# Ceph - scalable distributed file system
+#
+# Copyright (C) 2022 Red Hat, Inc.
+#
+# This is free software; you can redistribute it and/or
+# modify it under the terms of the GNU Lesser General Public
+# License version 2.1, as published by the Free Software
+# Foundation.  See file COPYING.
+
+# Suggested recovery sequence (for single MDS cluster):
+#
+# 1) Unmount all clients.
+#
+# 2) Flush the journal (if possible):
+#
+#    ceph tell mds.<fs_name>:0 flush journal
+#
+# 3) Fail the file system:
+#
+#    ceph fs fail <fs_name>
+#
+# 4a) Recover dentries from the journal. This will be a no-op if the MDS flushed the journal successfully:
+#
+#    cephfs-journal-tool --rank=<fs_name>:0 event recover_dentries summary
+#
+# 4b) If all good so far, reset the journal:
+#
+#    cephfs-journal-tool --rank=<fs_name>:0 journal reset
+#
+# 5) Run this tool to see list of damaged dentries:
+#
+#    python3 first-damage.py --memo run.1 <pool>
+#
+# 6) Optionally, remove them:
+#
+#    python3 first-damage.py --memo run.2 --remove <pool>
+#
+# Note: use --memo to specify a different file to save objects that have
+# already been traversed, for independent runs.
+#
+# This has the effect of removing that dentry from the snapshot or HEAD
+# (current hierarchy).  Note: the inode's linkage will be lost. The inode may
+# be recoverable in lost+found during a future data scan recovery.
+
+import argparse
+import logging
+import os
+import rados
+import re
+import sys
+import struct
+
+log = logging.getLogger("first-damage-traverse")
+
+MEMO = None
+REMOVE = False
+POOL = None
+NEXT_SNAP = None
+CONF = None
+
+DIR_PATTERN = re.compile(r'[0-9a-fA-F]{8,}\.[0-9a-fA-F]+')
+
+CACHE = set()
+
+def traverse(MEMO, ioctx):
+    for o in ioctx.list_objects():
+        if not DIR_PATTERN.fullmatch(o.key):
+            log.debug("skipping %s", o.key)
+            continue
+        elif o.key in CACHE:
+            log.debug("skipping previously examined object %s", o.key)
+            continue
+        log.info("examining: %s", o.key)
+
+        with rados.ReadOpCtx() as rctx:
+            it = ioctx.get_omap_vals(rctx, None, None, 100000)[0]
+            ioctx.operate_read_op(rctx, o.key)
+            for (dnk, val) in it:
+                log.debug('\t%s', dnk)
+                (first,) = struct.unpack('<I', val[:4])
+                if first > NEXT_SNAP:
+                    log.warn(f"found {o.key}:{dnk} first (0x{first:x}) > NEXT_SNAP (0x{NEXT_SNAP:x})")
```

Comment:
log.warn is deprecated in favor of log.warning since python 3.3

I spotted this while using the tool
```
$ python3 ../src/tools/cephfs/first-damage.py --memo /tmp/memo cephfs.a.meta --conf=/home/rraja/git/ceph/build/ceph.conf
/home/rraja/git/ceph/build/../src/tools/cephfs/first-damage.py:82: DeprecationWarning: The 'warn' method is deprecated, use 'warning' instead
  log.warn(f"found {o.key}:{dnk} first (0x{first:x}) > NEXT_SNAP (0x{NEXT_SNAP:x})")
```

Reply:
fixed